src/i2c.rs

```diff
@@ -7,186 +7,326 @@
+//!
+//! // Finalise expectations
+//! i2c.done();
 //! ```
 
-use std::io::Read;
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
 
 use hal::blocking::i2c;
 
 use error::MockError;
 
-const WRITE_BUF_SIZE: usize = 64;
-
-/// Mock I²C implementation
-///
-/// The main idea behind this is that you set the data that will be read back
-/// by a read call in advance. Then the driver can read from the mock device,
-/// and will get exactly the data you prepared.
-///
-/// See the usage section in the module level docs for an example.
-pub struct Mock<'a> {
-    data: &'a [u8],
-    address: Option<u8>,
-    buf: [u8; WRITE_BUF_SIZE],
-    buf_bytes_written: usize,
+/// I2C Transaction modes
+#[derive(Clone, Debug, PartialEq)]
+pub enum Mode {
+    Write,
+    Read,
+    WriteRead,
+}
+
+/// I2C Transaction type
+/// Models an I2C read or write
+#[derive(Clone, Debug, PartialEq)]
+pub struct Transaction {
+    expected_mode: Mode,
+    expected_addr: u8,
+    expected_data: Vec<u8>,
+    response_data: Vec<u8>,
+}
+
+impl Transaction {
+    /// Create a Write transaction
+    pub fn write(addr: u8, expected: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Write,
+            expected_addr: addr,
+            expected_data: expected,
+            response_data: Vec::new(),
+        }
+    }
+
+    /// Create a Read transaction
+    pub fn read(addr: u8, response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::Read,
+            expected_addr: addr,
+            expected_data: Vec::new(),
+            response_data: response,
+        }
+    }
+
+    /// Create a WriteRead transaction
+    pub fn write_read(addr: u8, expected: Vec<u8>, response: Vec<u8>) -> Transaction {
+        Transaction {
+            expected_mode: Mode::WriteRead,
+            expected_addr: addr,
+            expected_data: expected,
+            response_data: response,
+        }
+    }
+}
+
+/// Mock I2C implementation
+/// This supports the specification and evaluation of expectations to allow automated testing of I2C based drivers.
+/// Mismatches between expectations will cause runtime assertions to assist in locating the source of the fault.
+pub struct Mock {
+    expected: Arc<Mutex<VecDeque<Transaction>>>,
 }
 
-impl<'a> Mock<'a> {
+impl Mock {
     pub fn new() -> Self {
         Mock {
-            data: &[],
-            address: None,
-            buf: [0; WRITE_BUF_SIZE],
-            buf_bytes_written: 0,
+            expected: Arc::new(Mutex::new(VecDeque::new())),
         }
     }
 
-    /// Set data that will be read by `read()`.
-    pub fn set_read_data(&mut self, data: &'a [u8]) {
-        self.data = data;
+    /// Set expectations on the I2C interface
+    /// This is a list of I2C transactions to be executed in order
+    /// Note that setting this will overwrite any existing expectations
+    pub fn expect(&mut self, expected: Vec<Transaction>) {
```

Comment:
At least here you could easily change the `expected` type to `&[Transaction]`, allowing the user to specify a static array.